File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.js
function solution(players, m, k) {
    const server = []
    let count = 0;
    
    for(let i = 0; i < players.length; i++) {
        const n = server.reduce((acc, cur) => cur.count + acc, 0)
        const player = players[i];
        
        if(player >= (n + 1) * m) {
            const need = Math.floor((player / m)) - n  
            server.push({exit: i + k - 1, count: need})
            count += need
        }

        while(server[0] && server[0].exit === i) server.shift()
    }
    
    return count
}

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.*;

class Solution {
    public int solution(int[] players, int m, int k) {
        int answer = 0;
        int count = 0;
        List<Server> server = new ArrayList<>();
        
        for(int i = 0; i < players.length; i++) {
            
            int n = 0;
            
            for (Server s : server) {
                n += s.count;
            }
        
            int player = players[i];
                
            if(player >= (n+1) * m) {
                int need = (player / m) - n;
                server.add(new Server(i + k - 1, need));
                count += need;
            }
            
            while(!server.isEmpty() && server.get(0).exit == i) {
                server.remove(0);
            }
            
        }
        
        return count;
    }
    
    static class Server {
        int exit;
        int count; 
        
        Server(int exit, int count) {
            this.exit = exit;
            this.count = count;
        }
    }
}